R/analysis.R

```diff
@@ -181,6 +181,188 @@ netAnalysis_contribution <- function(object, signaling, signaling.name = NULL, s
+#' @importFrom dplyr select
+#' @importFrom ggplot2 ggplot geom_bar aes coord_flip scale_x_discrete element_text theme ggtitle
+#' @importFrom cowplot ggdraw draw_label plot_grid
+#'
+#' @return
+#' @export
+#' 
+#' @note This function is different from `netAnalysis_contribution` in two ways: 1) no need to zoom into a specific pathway; 2) do not count the relative prob. Overall, this function makes the resulted contribution comparable across sources and targets. 
+#' @examples
+netAnalysis_contribution_allLR <- function(object, signaling = NULL, sources.use = NULL, targets.use = NULL,
+                                           width = 0.1, vertex.receiver = NULL, thresh = 0.05, return.data = FALSE,
+                                           x.rotation = 0, title = "Contribution of each L-R pair",
+                                           font.size = 10, font.size.title = 10) {
+  if (is.null(signaling)) {signaling <- object@netP$pathways}
+  pairLR <- searchPair(signaling = signaling, pairLR.use = object@LR$LRsig, key = "pathway_name", matching.exact = T, pair.only = T)
+  pair.name.use = select(object@DB$interaction[rownames(pairLR),],"interaction_name_2")
+
+  net <- object@net
+  pairLR.use.name <- dimnames(net$prob)[[3]]
+  pairLR.name <- intersect(rownames(pairLR), pairLR.use.name)
+  pairLR <- pairLR[pairLR.name, ]
+  prob <- net$prob
+  pval <- net$pval
+  
+  prob[pval > thresh] <- 0
+  
+  if (!is.null(sources.use)) {
+    if (is.character(sources.use)) {
+      if (all(sources.use %in% dimnames(prob)[[1]])) {
+        sources.use <- match(sources.use, dimnames(prob)[[1]])
+      } else {
+        stop("The input `sources.use` should be cell group names or a numerical vector!")
+      }
+    }
+    idx.t <- setdiff(1:nrow(prob), sources.use)
+    prob[idx.t, , ] <- 0
+  }
+  
+  if (!is.null(targets.use)) {
+    if (is.character(targets.use)) {
+      if (all(targets.use %in% dimnames(prob)[[1]])) {
+        targets.use <- match(targets.use, dimnames(prob)[[2]])
+      } else {
+        stop("The input `targets.use` should be cell group names or a numerical vector!")
+      }
+    }
+    idx.t <- setdiff(1:nrow(prob), targets.use)
+    prob[ ,idx.t, ] <- 0
+  }
+  
+  if (length(pairLR.name) > 1) {
+    pairLR.name.use <- pairLR.name[apply(prob[,,pairLR.name], 3, sum) != 0]
+  } else {
+    pairLR.name.use <- pairLR.name[sum(prob[,,pairLR.name]) != 0]
+  }
+  
+  
+  if (length(pairLR.name.use) == 0) {
+    stop(paste0('There is no significant communication'))
+  } else {
+    pairLR <- pairLR[pairLR.name.use,]
+  }
+  
+  prob <- prob[,,pairLR.name.use]
+  
+  if (length(dim(prob)) == 2) {
+    prob <- replicate(1, prob, simplify="array")
+    dimnames(prob)[3] <- pairLR.name.use
+  }
+  #prob <-(prob-min(prob))/(max(prob)-min(prob)) # this is the relative prob
+  
+  if (is.null(vertex.receiver)) {
+    pSum <- apply(prob, 3, function(pmat) {sum(pmat) / length(unique(c(sources.use, targets.use)))})
```

Comment:
With more sources/targets, the `sum` will always be higher; so the number of sources/targets should be somehow adjusted. Here, I just divide, but there could be a better solution.